### PR TITLE
trunc_normal_ low precision fix 

### PR DESCRIPTION
**Pull request by albanD**
Fixes https://github.com/pytorch/pytorch/issues/145498

A bulk of the change is adding a lot of test for dtypes, devices, etc
Otherwise, we actually stop using low precision uniform and go with uniform + resampling.

The more detailed explanation for why we want to move away form the old implementation and in particular erfinv is:
 When std=0.002 and a=-2, b=2, the CDF bounds become:
  - l = norm_cdf(-1000) = 0.0 (underflows to exactly zero in fp64)
  - u = norm_cdf(1000) = 1.0

  So the built-in calls uniform_(-1.0, 1.0) then erfinv_(). The problem: erfinv(x) → ±∞ as x → ±1. In fp32, values very close to ±1 produce enormous outputs.
  After scaling by std * sqrt(2) ≈ 0.00283, most are fine, but a few extreme ones survive and get clamped to ±2.0 — which is 1000 standard deviations from the
   mean.

  The damage:
| Metric | built-in | rejection | expected |
|---|---|---|---|
| std | 0.002029 | 0.002000 | 0.002000 |
| kurtosis | 28,122 | 3.00 | ~3.0 |
| min | -2.0 (clamp) | -0.011 | ~-0.011 |
| max | 0.011 | 0.011 | ~0.011 |
| elements at ±2.0 | 2 | 0 | 0 |

  The built-in's std is inflated (0.00203 vs 0.002) and its kurtosis is off by 4 orders of magnitude — just 2 elements clamped at -2.0 (1000σ outliers) are
  enough to destroy the 4th moment of 67M samples.

  The quantiles look OK because the outliers only affect the extreme tails and moments. But for any application sensitive to variance or tail behavior (weight
   initialization for training stability), those clamped values are pure noise.

  The rejection-sampling approach doesn't have this problem — normal_() uses cuRAND's native normal generator (Box-Muller), which doesn't suffer from erfinv's
   precision collapse near ±1.

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: torch/nn/init.py [hunk 1]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
+            # Cast bounds to tensor dtype so the rejection mask is consistent
+            # with the tensor's representable values.
+            lo = tensor.new_tensor(a).item()
+            hi = tensor.new_tensor(b).item()
+            tensor.normal_(mean, std, generator=generator)
+            while True:
+                mask = (tensor < lo) | (tensor > hi)
+                if not mask.any():
+                    break
+                tensor[mask] = torch.normal(
```
Comment (ngimel) [inline torch/nn/init.py]: why not torch.where here?

File: torch/nn/init.py [hunk 2]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
+                tensor.shape, dtype=torch.bool, device=tensor.device
+            )
+            while not accepted.all():
+                pending = ~accepted
+                count = pending.sum().item()
+                tensor[pending] = torch.empty(
+                    count, dtype=tensor.dtype, device=tensor.device,
+                ).uniform_(a, b, generator=generator)
+                log_pdf = -0.5 * ((tensor[pending] - mean) / std) ** 2
+                accepted[pending] = torch.rand(
```
Comment (Skylion007) [inline torch/nn/init.py]: Can use torch.rand_like directly

File: torch/nn/init.py [hunk 3]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
+                    size=(mask.sum().item(),),
+                    dtype=tensor.dtype,
+                    device=tensor.device,
+                    generator=generator,
+                )
+        else:
+            mode = max(a, min(mean, b))
+            log_peak = -0.5 * ((mode - mean) / std) ** 2
+
+            accepted = torch.zeros(
```
Comment (Skylion007) [inline torch/nn/init.py]: Any reason not zeros_like? Do we care about making the output strides contiguous? We can specify that explicitly if so...

File: torch/nn/init.py [hunk 4]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
+                tensor.shape, dtype=torch.bool, device=tensor.device
+            )
+            while not accepted.all():
+                pending = ~accepted
+                count = pending.sum().item()
```
Comment (Skylion007) [inline torch/nn/init.py]: Does count actually need to be roundtripped to host explicitly? Or is this just to cache it as a cpu types? Seems fine to leave it as a Tensor no item call necessary?
```suggestion
                count = pending.sum()
```

File: torch/nn/init.py [hunk 5]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
+                tensor[pending] = torch.empty(
```
Comment (Skylion007) [inline torch/nn/init.py]: Can cache the view as a temp variable with t_pending = tensor[pending]...
```suggestion
                t_pending = tensor[pending] = torch.empty(

File: torch/nn/init.py [hunk 6]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
+                    count, dtype=tensor.dtype, device=tensor.device,
+                ).uniform_(a, b, generator=generator)
+                log_pdf = -0.5 * ((tensor[pending] - mean) / std) ** 2
+                accepted[pending] = torch.rand(
+                    count, dtype=tensor.dtype, device=tensor.device,
+                ).log() <= (log_pdf - log_peak)
```
Comment (Skylion007) [inline torch/nn/init.py]: Pretty sure PyTorch has inplace log. Use that: https://docs.pytorch.org/docs/stable/generated/torch.Tensor.log_.html
```suggestion
                ).log_() <= (log_pdf - log_peak)

File: torch/nn/init.py [hunk 7]
```diff
@@ -91,9 +91,7 @@ def _no_grad_trunc_normal_(
     b: float,
     generator: torch.Generator | None = None,
 ) -> Tensor:
-    # Method based on https://people.sc.fsu.edu/~jburkardt/presentations/truncated_normal.pdf
```
Comment (Skylion007) [inline torch/nn/init.py:94]: Why remove these useful comments?

File: torch/nn/init.py [hunk 8]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
+            )
+            while not accepted.all():
+                pending = ~accepted
+                count = pending.sum().item()
+                tensor[pending] = torch.empty(
+                    count, dtype=tensor.dtype, device=tensor.device,
+                ).uniform_(a, b, generator=generator)
+                log_pdf = -0.5 * ((tensor[pending] - mean) / std) ** 2
+                accepted[pending] = torch.rand(
+                    count, dtype=tensor.dtype, device=tensor.device,
```
Comment (Skylion007) [inline torch/nn/init.py]:
```suggestion
                accepted[pending] = torch.rand_like(t_pending
```

File: torch/nn/init.py [hunk 9]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
+                    dtype=tensor.dtype,
+                    device=tensor.device,
+                    generator=generator,
+                )
+        else:
+            mode = max(a, min(mean, b))
+            log_peak = -0.5 * ((mode - mean) / std) ** 2
+
+            accepted = torch.zeros(
+                tensor.shape, dtype=torch.bool, device=tensor.device
```
Comment (Skylion007) [inline torch/nn/init.py]:
```suggestion
            accepted = torch.zeros_like(tensor, dtype=torch.bool)

File: torch/nn/init.py (hunk 1 above)
Comment (Skylion007) [reply torch/nn/init.py]: or where_ or masked_select, etc...

Comment (Skylion007) [reply torch/nn/init.py]:
```suggestion
                tensor[mask].normal_(
```

File: torch/nn/init.py [hunk 10]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
+                if not mask.any():
+                    break
+                tensor[mask] = torch.normal(
+                    mean,
+                    std,
+                    size=(mask.sum().item(),),
+                    dtype=tensor.dtype,
+                    device=tensor.device,
+                    generator=generator,
+                )
```
Comment (Skylion007) [inline torch/nn/init.py]:
```suggestion
                tensor[mask].normal_(
                    mean,
                    std,
                    generator=generator,
                )

Comment (Skylion007) [reply torch/nn/init.py]: You can just call normal_ on the view inplace, can't you?

File: torch/nn/init.py [hunk 11]
```diff
@@ -104,26 +102,44 @@ def norm_cdf(x: float) -> float:
...
-
-        # Clamp to ensure it's in the proper range
-        tensor.clamp_(min=a, max=b)
+        p = norm_cdf((b - mean) / std) - norm_cdf((a - mean) / std)
+
+        if p > 0.3:
+            # Cast bounds to tensor dtype so the rejection mask is consistent
+            # with the tensor's representable values.
+            lo = tensor.new_tensor(a).item()
+            hi = tensor.new_tensor(b).item()
```
Comment (Skylion007) [inline torch/nn/init.py]:
```suggestion
            lo = tensor.new_tensor(a)
            hi = tensor.new_tensor(b)
```
Why do they need to be a Python scalar? Let it live on the same device as the tensor to minimize d2h, h2d syncs.

File: torch/nn/init.py (hunk 7 above)
Comment (ngimel) [reply torch/nn/init.py:94]: Because it's not longer based on that (very numerically bad) method

File: torch/nn/init.py (hunk 4 above)
Comment (ngimel) [inline torch/nn/init.py]: `count` is needed on cpu to create tensors, but we can remove one cpu sync 
```suggestion
            count = tensor.numel()
            while count > 0:
                pending = ~accepted
                ...
                count = tensor.numel() - accepted.sum().item()
```

Comment (ngimel) [reply torch/nn/init.py]: Although this implementation with boolean mask indexing seems pretty hopeless wrt host syncs (the upper branch, if rewritten with where instead of boolean mask would be better, except for once-per-iteration sync for ending condition

File: torch/nn/init.py (hunk 11 above)
Comment (albanD) [reply torch/nn/init.py]: It's actually even better to never send it to the device. Updated.

File: torch/nn/init.py (hunk 10 above)
Comment (albanD) [reply torch/nn/init.py]: No, mask advanced indexing does not return a view.

File: torch/nn/init.py (hunk 1 above)
Comment (albanD) [reply torch/nn/init.py]: torch.where doesn't work because we need to change the tensor inplace.
where_ doesn't exist.
masked_select wouldn't provide us with a view, so we wouldn't be able to change the result inplace.

I changed it to masked_scatter_, but I suspect this is exactly equivalent to what the advanced indexing was doing.

File: torch/nn/init.py (hunk 4 above)
Comment (albanD) [reply torch/nn/init.py]: Updated in a similar way

**Review by ngimel — APPROVED** [review]

**Comment by albanD** [comment]
@claude verify all the previous failures are fixed
